Feature/update example data code

vgro opened the pull request:
This PR moves the recipes for creating dummy data from the documentation to the example_data folder as code that can be run directly to reproduce the data.
I also added a page that describes the example data; at the moment this is a bit cryptic because I wasn't sure what we want this to look like. I feel like there should be some info about the assumptions available in the user documentation. Please give feedback and suggestions on what this could look like.
Finally, I updated the climate data related documentation as a more general resource with lots of links to the code.

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

jacobcook1995 reviewed (approved):
This looks sensible to me! In terms of the soil and litter dummy data I don't think it makes sense to talk about them much in the user documentation as they really are just random values to test that the model actually runs sensibly. Obviously, this would change if I e.g. started taking data from soil maps for the example data.

vgro commented:
> This looks sensible to me! In terms of the soil and litter dummy data I don't think it makes sense to talk about them much in the user documentation as they really are just random values to test that the model actually runs sensibly. Obviously, this would change if I e.g. started taking data from soil maps for the example data.

Ok, that makes sense. Maybe I just make it a bit more general and easy to read :-)

davidorme commented:
I've pushed a bunch of changes here to add some structure to the directory and to synchronize some common elements. It fails because I haven't updated the test structure yet. I'm waiting on `plants` being merged first but then will sort out fixing the tests and documenting the example use.

davidorme commented:
I think this is now ready for review and in a state to merge if everyone is happy.  It's probably useful to start by reading 
[docs/source/virtual_rainforest/example_data.md](https://github.com/ImperialCollegeLondon/virtual_rainforest/pull/339/files#diff-77aa7ee7797ab9ac38a9a06e53a0d2d04e2a0da0d004bc93760b92d4db46a64b) since that describes the changes in some detail. 

* I've moved from `dummy` to `example` throughout - none of the inputs are `dummy` for missing models anymore.
* I've organised the files within the folder into `generation_scripts`, `data` and `config` subdirectories to make it clearer what all the components are.
* I've updated the example description to include tables of variables for each component and added expandable links to the generation scripts and config TOML.

I've also expanded the config file path resolution a bit to cope with different path combinations. There is some minor `Config` method renaming here: we now _collect_ config TOML files (not resolve them) and we _resolve_ config file paths (not fix up). @alexdewar - these are changes to the system you put in place. I think it all makes sense and resolves as I expect, but it is broken on Windows + Python 3.9 for reasons I can't fathom.

@vgro - I can't seem to add you to review the PR. I don't know how GH handles multi-author PRs!

jacobcook1995 commented on docs/source/virtual_rainforest/example_data.md:
```diff
@@ -0,0 +1,340 @@
+
+This code creates a set of plausible values for which the
+{mod}`~virtual_rainforest.models.soil.soil_model` absolutely has to function sensibly
+for.Descriptions of the soil pools can be found [here](./soil/soil_details.md).
```
space needed between "for." and "Descriptions

jacobcook1995 commented on tests/core/test_config.py:
```diff
@@ -731,104 +729,83 @@ def test_Config_export_config(caplog, shared_datadir, auto, expected_log_entries
+        #     {"file": "file.txt", "var_name": "my_path"},
+        #     {"file": "file.txt", "var_name": "my_path"},
+        #     id="variable_not_list",
+        # ),
```
I guess either comment these back in or delete them

vgro replied:
This reminds me of the problem we had with multiple config files (one from each model) has this been resolved?

jacobcook1995 replied:
Nope that issue (#234) hasn't been resolved, multiple config files are completely, but all of the instructions of where to load data from have to be contained in a single config file

jacobcook1995 reviewed (approved):
LGTM, just had a few minor comments.

vgro commented on docs/source/virtual_rainforest/example_data.md:
```diff
@@ -0,0 +1,340 @@
+  - unitless
+  - XYT
+* - atmospheric pressure
+  - `precipitation`
```
I think you mixed precipitation and pressure detail up here? 
```suggestion
  - `atmospheric_pressure_ref`
```

vgro commented on docs/source/virtual_rainforest/example_data.md:
```diff
@@ -0,0 +1,340 @@
+  - kPa
+  - XYT
+* - precipitation
+  - `atmospheric_pressure_ref`
```
```suggestion
  - `precipitation`
```

vgro commented on docs/source/virtual_rainforest/example_data.md:
```diff
@@ -0,0 +1,340 @@
+  - `pH` 
+  - unitless
+  - XY
+* - Bulk soul density
```
```suggestion
* - Bulk soil density
```

vgro commented on docs/source/virtual_rainforest/example_data.md:
```diff
@@ -0,0 +1,340 @@
+  - m
+  - C
+* - Photosynthetic photon flux density
+  - photosynthetic_photon_flux_density
```
```suggestion
  - `photosynthetic_photon_flux_density`
```

vgro commented on virtual_rainforest/example_data/generation_scripts/elevation_example_data.py:
```diff
@@ -0,0 +1,75 @@
+from virtual_rainforest.example_data.generation_scripts.common import cell_displacements
+
+# # Load DEM in 30m resolution
+# original_data = requests.get(
```
What happened here, why did you outcomment this part?

davidorme replied:
It seemed like a lot of work to download and preprocess a fairly large file to generate 81 numbers, particularly when those numbers aren't currently meaningfully aligned to reality 😄 

At some point, this will become relevant again, but to make it reproducible and easy to re-run here and now, it seemed easier to retire the recipe and just provide the values.

vgro commented:
> I think this is now ready for review and in a state to merge if everyone is happy. It's probably useful to start by reading [docs/source/virtual_rainforest/example_data.md](https://github.com/ImperialCollegeLondon/virtual_rainforest/pull/339/files#diff-77aa7ee7797ab9ac38a9a06e53a0d2d04e2a0da0d004bc93760b92d4db46a64b) since that describes the changes in some detail.
> 
>     * I've moved from `dummy` to `example` throughout - none of the inputs are `dummy` for missing models anymore.
> 
>     * I've organised the files within the folder into `generation_scripts`, `data` and `config` subdirectories to make it clearer what all the components are.
> 
>     * I've updated the example description to include tables of variables for each component and added expandable links to the generation scripts and config TOML.
> 
> 
> I've also expanded the config file path resolution a bit to cope with different path combinations. There is some minor `Config` method renaming here: we now _collect_ config TOML files (not resolve them) and we _resolve_ config file paths (not fix up). @alexdewar - these are changes to the system you put in place. I think it all makes sense and resolves as I expect, but it is broken on Windows + Python 3.9 for reasons I can't fathom.
> 
> @vgro - I can't seem to add you to review the PR. I don't know how GH handles multi-author PRs!

This might be because I create the PR. I added a few minor comments but it looks overall good to me
